src/Extension/ExtensionFactory.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OAT\DependencyResolver\Extension;
+
+use OAT\DependencyResolver\Extension\Entity\Extension;
+use OAT\DependencyResolver\Extension\Exception\NotMappedException;
+
+class ExtensionFactory
+{
+    /** @var array */
+    private $extensionMap = [];
+
+    public function __construct(array $extensionMap)
+    {
+        $this->extensionMap = $extensionMap;
+    }
+
+    /**
+     * @param string $extensionName
+     * @param string $branch
+     *
+     * @return Extension
+     * @throws NotMappedException
+     */
+    public function create(string $extensionName, string $branch = Extension::DEFAULT_BRANCH): Extension
+    {
+        if (!isset($this->extensionMap[$extensionName])) {
+            $extensionName = $this->findExtensionNameFromRepositoryName($extensionName);
+        }
+
+        return new Extension($extensionName, $this->extensionMap[$extensionName], $branch);
+    }
+
+    /**
+     * Tries to find an extension corresponding to the supposed repository name.
+     *
+     * @param string $extensionName
+     *
+     * @return string
+     * @throws NotMappedException when the name can not be mapped to any extension.
+     */
+    private function findExtensionNameFromRepositoryName(string $extensionName)
```

Comment:
I don't really understand what is your intention with this method. If you cannot find it by index, you try to find it by value? This looks weird for me. If you cannot find it by index, you should throw the exception right away.

Reply:
Refactored with explicit extenstion-name or repository-name options => removed.